### ISSUE TITLE
"UnhandledException : Object reference not set to an instance of an object." in Run-TestsInNavContainer

### ISSUE THREAD

**Issue by ralf-escher** [microsoft/navcontainerhelper#906]
Hi!

Based on the scripts of @freddydk I created an Azure Pipleline which runs on a virtual Azure Machine as Agent.
I didn't changed the scripts much and I could run everything pretty well for a small app with a very simple test codeunit.

Then I tried the same with our (quite bigger) app and compiling, publishing works also well.
If I try to run my test codeunits (mostly integration tests splitted into six codeunits), I always get the same following error (no matter which of my test codeunits, one or many).
_Remark: If I run the tests manually in the webclient (on the virtual Azure machine) all is good._

_If I check the page "AL Test Tool" after the failed run, the first test codeunit (which is mentioned in the following error) has succeeded all of its routines, but stopped afterwards for the next codeunit._

```
  Codeunit 5140501 BET AUT Test Basic UnhandledException : Object reference not set to an instance of an object.
at <ScriptBlock>, C:\ProgramData\NavContainerHelper\Extensions\be-industry-build\PsTestTool-5\ClientContext.ps1: line 115
at <ScriptBlock>, <No file>: line 79
Current Interaction: Microsoft.Dynamics.Framework.UI.Client.Interactions.InvokeActionInteraction
Time spend: 12 seconds
ClientSession State is InError (Wait time 12 seconds)
at AwaitState, C:\ProgramData\NavContainerHelper\Extensions\be-industry-build\PsTestTool-5\ClientContext.ps1: line 198
at InvokeInteraction, C:\ProgramData\NavContainerHelper\Extensions\be-industry-build\PsTestTool-5\ClientContext.ps1: line 216
at InvokeAction, C:\ProgramData\NavContainerHelper\Extensions\be-industry-build\PsTestTool-5\ClientContext.ps1: line 421
at Run-Tests, C:\ProgramData\NavContainerHelper\Extensions\be-industry-build\PsTestTool-5\PsTestFunctions.ps1: line 399
at <ScriptBlock>, <No file>: line 53
##[error]ClientSession State is InError (Wait time 12 seconds)
In C:\Program Files\WindowsPowerShell\Modules\navcontainerhelper\0.6.5.0\AppHandling\Run-TestsInNavContainer.ps1:396 
Zeichen:17
+                 throw $_.Exception.Message
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~

    + CategoryInfo          : OperationStopped: (ClientSession S...ime 12 seconds):String) [], RuntimeException

    + FullyQualifiedErrorId : ClientSession State is InError (Wait time 12 seconds)
##[error]PowerShell exited with code '1'.
Finishing: Run Tests
```

Since the simple app works I reduced the number of test codeunits to one simple codeunit with three functions, but I got the same result.

Then I tried many configurations (start-sleep 180, isolation hyperv, more ram for the container, switched the azure machine with more RAM,...) but couldn't find the problem.

I also applied your following recommendation:
https://community.dynamics.com/business/b/freddysblog/posts/running-tests-in-containers-1945258721

-**debugMode** didn't showed something new
-**usePublicWebBaseUrl** didn't change anything
-Eventlog of the container reveals some stuff:

lots of `Message Object metadata load attempted with no thread app group context defined -- Type:Page; Id:<xxxx>`

this one shows three errors one after the other:
`Message Fatal error during metadata load -- Type:Query; Id:20000; MetadataGroup:18; MetadataLoaded:False; CLRLoaded:False; ThreadGroup:0
` 

and

```
Message (NavNCLMetadataObjectNotFoundException): RootException: NavNCLMetadataObjectNotFoundException
Object of type Table with ID 21 could not be found.
ExceptionStackTrace:
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObjectInternal(ApplicationObjectId appObjectId, Boolean requireCompiled, Boolean onlySearchCache, Int32 appGroupId)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObject(ApplicationObjectId appObjectId, Int32 appGroupId, Boolean requireCompiled, Boolean onlySearchCache)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaTableById(Int32 tableId, Int32 appGroupId, Boolean requireCompiled, Boolean onlySearchCache)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQueryDataItem.ResolveMetaTable(NavAppGroup appGroup)
   at Microsoft.Dynamics.Nav.Runtime.AppGroupAwareLookup`1.GetReferenceUsingSpecificAppGroup(NavAppGroup appGroup)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.CreateTableFiltersAndMarksFromDataItemFieldFilters(NCLMetaQueryDataItem dataItemBeingCreated, List`1 fieldFilters, Boolean allowExpressionEvaluation, Boolean& tableFilterHasVariables)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.<>c__DisplayClass67_1.<CreateDataItemsFromDesignMetadata>b__1(NCLMetaQueryDataItem dataItemBeingCreated)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQueryDataItem..ctor(NCLMetaQueryDefinition parentQueryDefinition, DataItemLinkType dataItemLinkType, Int32 tableNo, DataLockState lockState, String name, Func`2 queryColumnsCreator, Func`2 tableFiltersAndMarksCreator, Func`2 dataItemLinksCreator, List`1 metaQueryFieldFilters, Int32 appGroup)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.CreateDataItemsFromDesignMetadata(NCLMetaQueryDefinition parentQueryDefinition, MetaQuery metaQuery, Boolean allowExpressionEvaluation, Boolean& tableFilterHasVariables)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.<>c__DisplayClass68_1.<ParseMetadata>b__0(NCLMetaQueryDefinition queryDefinitionBeingCreated)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQueryDefinition..ctor(NCLMetaQueryDataItem parentDataItem, Int32 topRows, Func`2 dataItemsCreator, Func`2 orderBysCreator)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.ParseMetadata(MetaQuery metaQuery, Boolean allowExpressionEvaluation)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.LoadMetadata()
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaApplicationObject.Populate()
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObjectInternal(ApplicationObjectId appObjectId, Boolean requireCompiled, Boolean onlySearchCache, Int32 appGroupId)
CallerStackTrace:
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObjectInternal(ApplicationObjectId appObjectId, Boolean requireCompiled, Boolean onlySearchCache, Int32 appGroupId)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObject(ApplicationObjectId appObjectId, Int32 appGroupId, Boolean requireCompiled, Boolean onlySearchCache)
   at Microsoft.Dynamics.Nav.Runtime.ApiWebServiceEntryHelper.GetAllVersions(ObjectType objectType, Int32 objectId, PackageId packageId, NavAppGroup group, HashSet`1 paths, Boolean throwOnEndpointDuplicate)
   at Microsoft.Dynamics.Nav.Runtime.ApiWebServiceEntryHelper.GetEntriesFromNavApp(NavTenant tenant, HashSet`1 paths)
   at Microsoft.Dynamics.Nav.Runtime.WebServiceHelper.GetApiWebServices(NavTenant tenant)
   at Microsoft.Dynamics.Nav.Service.Api.ApiRouteManager.ConfigureCommonRoutesForAllTenants()
   at Microsoft.Dynamics.Nav.Service.Api.GraphApiHost.ConfigureHost(WebApiHost context, IAppBuilder builder, Category telemetryCategory, AuthenticationSchemes authenticationScheme, DelegatingHandler handler)
   at Microsoft.Dynamics.Nav.Service.WebApi.WebApiHost.<Open>b__31_0(IAppBuilder builder)
   at Microsoft.Owin.Hosting.Engine.HostingEngine.Start(StartContext context)
   at Microsoft.Dynamics.Nav.Service.WebApi.WebApiHost.Open(String machineName, ServerUserSettings settings)
   at Microsoft.Dynamics.Nav.WindowsServices.NavServerWindowsService.StartOwinServices(ServerUserSettings settings)
   at Microsoft.Dynamics.Nav.WindowsServices.NavServerWindowsService.StartAllServices(ServerUserSettings settings)
   at Microsoft.Dynamics.Nav.WindowsServices.NavServerWindowsService.Start(String commandLineServiceInstanceName)
   at System.ServiceProcess.ServiceBase.ServiceQueuedMainCallback(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

and finally

```
Server instance: BC
Category: Extensions
ClientSessionId: 00000000-0000-0000-0000-000000000000
ClientActivityId: b7fc239c-a0aa-4317-9369-24f4e2524532
ServerSessionUniqueId: 00000000-0000-0000-0000-000000000000
ServerActivityId: 120a698b-d581-4af5-aa8e-37294ec1b084
EventTime: 03/23/2020 09:30:03
Message (NavNCLMetadataObjectNotFoundException): This message had personal data removed. The original may still be in transient telemetry. Find it using the transientTelemetryId.
RootException: NavNCLMetadataObjectNotFoundException
Message not shown because the NavBaseException(string, Exception, bool) constructor was used.
ExceptionStackTrace:
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObjectInternal(ApplicationObjectId appObjectId, Boolean requireCompiled, Boolean onlySearchCache, Int32 appGroupId)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObject(ApplicationObjectId appObjectId, Int32 appGroupId, Boolean requireCompiled, Boolean onlySearchCache)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaTableById(Int32 tableId, Int32 appGroupId, Boolean requireCompiled, Boolean onlySearchCache)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQueryDataItem.ResolveMetaTable(NavAppGroup appGroup)
   at Microsoft.Dynamics.Nav.Runtime.AppGroupAwareLookup`1.GetReferenceUsingSpecificAppGroup(NavAppGroup appGroup)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.CreateTableFiltersAndMarksFromDataItemFieldFilters(NCLMetaQueryDataItem dataItemBeingCreated, List`1 fieldFilters, Boolean allowExpressionEvaluation, Boolean& tableFilterHasVariables)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.<>c__DisplayClass67_1.<CreateDataItemsFromDesignMetadata>b__1(NCLMetaQueryDataItem dataItemBeingCreated)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQueryDataItem..ctor(NCLMetaQueryDefinition parentQueryDefinition, DataItemLinkType dataItemLinkType, Int32 tableNo, DataLockState lockState, String name, Func`2 queryColumnsCreator, Func`2 tableFiltersAndMarksCreator, Func`2 dataItemLinksCreator, List`1 metaQueryFieldFilters, Int32 appGroup)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.CreateDataItemsFromDesignMetadata(NCLMetaQueryDefinition parentQueryDefinition, MetaQuery metaQuery, Boolean allowExpressionEvaluation, Boolean& tableFilterHasVariables)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.<>c__DisplayClass68_1.<ParseMetadata>b__0(NCLMetaQueryDefinition queryDefinitionBeingCreated)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQueryDefinition..ctor(NCLMetaQueryDataItem parentDataItem, Int32 topRows, Func`2 dataItemsCreator, Func`2 orderBysCreator)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.ParseMetadata(MetaQuery metaQuery, Boolean allowExpressionEvaluation)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaQuery.LoadMetadata()
   at Microsoft.Dynamics.Nav.Runtime.NCLMetaApplicationObject.Populate()
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObjectInternal(ApplicationObjectId appObjectId, Boolean requireCompiled, Boolean onlySearchCache, Int32 appGroupId)
CallerStackTrace:
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObjectInternal(ApplicationObjectId appObjectId, Boolean requireCompiled, Boolean onlySearchCache, Int32 appGroupId)
   at Microsoft.Dynamics.Nav.Runtime.NCLMetadata.GetMetaApplicationObject(ApplicationObjectId appObjectId, Int32 appGroupId, Boolean requireCompiled, Boolean onlySearchCache)
   at Microsoft.Dynamics.Nav.Runtime.ApiWebServiceEntryHelper.GetAllVersions(ObjectType objectType, Int32 objectId, PackageId packageId, NavAppGroup group, HashSet`1 paths, Boolean throwOnEndpointDuplicate)
   at Microsoft.Dynamics.Nav.Runtime.ApiWebServiceEntryHelper.GetEntriesFromNavApp(NavTenant tenant, HashSet`1 paths)
   at Microsoft.Dynamics.Nav.Runtime.WebServiceHelper.GetApiWebServices(NavTenant tenant)
   at Microsoft.Dynamics.Nav.Service.Api.ApiRouteManager.ConfigureCommonRoutesForAllTenants()
   at Microsoft.Dynamics.Nav.Service.Api.GraphApiHost.ConfigureHost(WebApiHost context, IAppBuilder builder, Category telemetryCategory, AuthenticationSchemes authenticationScheme, DelegatingHandler handler)
   at Microsoft.Dynamics.Nav.Service.WebApi.WebApiHost.<Open>b__31_0(IAppBuilder builder)
   at Microsoft.Owin.Hosting.Engine.HostingEngine.Start(StartContext context)
   at Microsoft.Dynamics.Nav.Service.WebApi.WebApiHost.Open(String machineName, ServerUserSettings settings)
   at Microsoft.Dynamics.Nav.WindowsServices.NavServerWindowsService.StartOwinServices(ServerUserSettings settings)
   at Microsoft.Dynamics.Nav.WindowsServices.NavServerWindowsService.StartAllServices(ServerUserSettings settings)
   at Microsoft.Dynamics.Nav.WindowsServices.NavServerWindowsService.Start(String commandLineServiceInstanceName)
   at System.ServiceProcess.ServiceBase.ServiceQueuedMainCallback(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**Comment by freddydk** [comment]
when you run the tests in the webclient - what is the time spend on running the tests of the first codeunit?
did you try to increase the timeout of run-tests?

**Comment by ralf-escher** [comment]
First test codeunit needs 17s.

**Comment by ralf-escher** [comment]
First test codeunit takes 17s.
Second test codeunit takes 2min 47s.

```
Run-TestsInNavContainer 
[timespan] $interactionTimeout = [timespan]::FromHours(24),
```
what is a useful value for `interactionTimeout`?

Should I try to trace down the problem with fiddler as a next step?

**Comment by freddydk** [comment]
Not sure fiddler would give anything - but you are more than welcome to try.
If you use fiddler, you need to use -connectFromHost (which you could try without fiddler first).
That will run the PS connecting to the container from the host (instead of from the inside)

**Comment by ralf-escher** [comment]
Could you give me an useful value for `interactionTimeout`?

The event log was also complaining (quite often) about slow execution:
```
Message Action completed successfully, but it took longer than the given threshold.
  Execution time: 1033 ms
  Threshold: 1000 ms
  Message: Long running SQL statement
```

**Comment by freddydk** [comment]
interactionTimeout is the timeout for pressing the "run tests" button, i.e. the time it takes for running the tests. The 24 hours was just set to make sure it didn't give a timeout - and it doesn't seem like you are running into any timeout issue.

**Comment by freddydk** [comment]
If you still have the issue, then if you can create a repro-scenario (something I can run on a clean Azure VM or like) and get the issue, then I can try to troubleshoot.

**Comment by ralf-escher** [comment]
Meanwhile I tried even more combinations and could run at least a few tests.
My first test codeunit failed due to my attempt to set the Workdate.
With the second codeunit I only could run a simple test at the moment.
Otherwise I always get an error.

The Event Log shows always the same error (for several different tests)
```
Object of type Table with ID 21 could not be found.
Message Fatal error during metadata load -- Type:Query; Id:5480; MetadataGroup:3; MetadataLoaded:False;
```
Found this similar issue:
https://github.com/microsoft/nav-docker/issues/435

Any hint to isolate this issue?

**Comment by freddydk** [comment]
Do you have an end 2 end script, which repros the error?
If you can share the script including the full output of everything, then I can try to reproduce the error.

**Comment by ralf-escher** [comment]
I have found out, that I can run the tests on the virtual Azure machine directly with the local container. (your script from https://community.dynamics.com/business/b/freddysblog/posts/running-tests-in-containers-1945258721 section "Running tests one codeunit at a time").
The result is the same, but this is way easier instead running the Azure DevOps Pipeline all the time.

```
ClientSession State is InError (Wait time 4 seconds)
at AwaitState, C:\ProgramData\NavContainerHelper\Extensions\be-industry-build\PsTestTool-5\ClientContext.ps1: line 198
```

Couldn't debug that PowerShell script. Any chance that I can get some better information from ClientContext.ps1 about the problem?

My current guess is that I have a problem with table locking and the routine runs into a time out.

**Comment by freddydk** [comment]
Do you have TaskScheduler turned on?

**Comment by freddydk** [comment]
I have just fixed a bug, which could be the cause of this.
this will be included in 0.6.5.7
if the issue persists, then the new version will also add a lot of extra debug info with -debugMode, so please reopen and include the new output here.
Thanks

**Comment by freddydk** [comment]
Shipped in 0.6.5.7
